client/CServerHandler.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * CServerHandler.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#pragma once
+
+#include "../lib/CStopWatch.h"
+
+#include "../lib/StartInfo.h"
+
+struct SharedMemory;
+class CConnection;
+class PlayerColor;
+struct StartInfo;
+
+class CMapInfo;
+struct ClientPlayer;
+struct CPack;
+struct CPackForLobby;
+class CClient;
+
+template<typename T> class CApplier;
+class CBaseForLobbyApply;
+
+// TODO: Add mutex so we can't set CONNECTION_CANCELLED if client already connected, but thread not setup yet
+enum class EClientState : ui8
+{
+	NONE = 0,
+	CONNECTING, // Trying to connect to server
+	CONNECTION_CANCELLED, // Connection cancelled by player, stop attempts to connect
+	LOBBY, // Client is connected to lobby
+	LOBBY_CAMPAIGN, // Client is on scenario bonus selection screen
+	STARTING, // Gameplay interfaces being created, we pause netpacks retrieving
+	GAMEPLAY, // In-game, used by some UI
+	DISCONNECTING // We disconnecting, drop all netpacks
+};
+
+class IServerAPI
+{
+protected:
+	virtual void sendLobbyPack(const CPackForLobby & pack) const = 0;
+
+public:
+	virtual ~IServerAPI() {}
+
+	virtual void sendClientConnecting() const = 0;
+	virtual void sendClientDisconnecting() = 0;
+	virtual void setCampaignState(std::shared_ptr<CCampaignState> newCampaign) = 0;
+	virtual void setCampaignMap(int mapId) const = 0;
+	virtual void setCampaignBonus(int bonusId) const = 0;
+	virtual void setMapInfo(std::shared_ptr<CMapInfo> to, std::shared_ptr<CMapGenOptions> mapGenOpts = {}) const = 0;
+	virtual void setPlayer(PlayerColor color) const = 0;
+	virtual void setPlayerOption(ui8 what, ui8 dir, PlayerColor player) const = 0;
+	virtual void setDifficulty(int to) const = 0;
+	virtual void setTurnLength(int npos) const = 0;
+	virtual void sendMessage(const std::string & txt) const = 0;
+	virtual void sendGuiAction(ui8 action) const = 0; // TODO: possibly get rid of it?
+	virtual void sendStartGame(bool allowOnlyAI = false) const = 0;
+};
+
+/// structure to handle running server and connecting to it
+class CServerHandler : public IServerAPI, public LobbyInfo
```

Comment:
One thing that made me wonder, why CServerHandler inherits from LobbyInfo.

Reply:
This is done for convenience. Efficiently I might only inherit ```LobbyState``` and put everything here, but right I can update it this way on client:
```
bool LobbyUpdateState::applyOnLobbyHandler(CServerHandler * handler)
{
	hostChanged = state.hostClientId != handler->hostClientId;
	static_cast<LobbyState &>(*handler) = state;
	return true;
}
```
It's just saved me time on editing that code compared to if I just applied each field separately. So I had UUID out since to ```LobbyInfo``` so let's say UUID not changes when new state applied on client.

Also you might argue I should just put whole ```LobbyState``` into separate variable on both server and client, but right now it's will only make half of lines of code longer without obvious benefit. Since I choose to iterate over existing code rather than try to rewrite everything from scratch there still plenty of other more important improvements needed than deciding on nice looking API for shared lobby state.

Reply:
Yeah it's would be the best way, but before that code quality need to be on the point where it's going to have somewhat clear API...